elasticsearch-transport/lib/elasticsearch/transport/transport/http/manticore.rb

```diff
@@ -43,97 +46,156 @@ module HTTP
+          def normalize_host(host)
+            case host
+              when URI
+                host.to_s
+              when String
+                URI.parse(host).to_s
+              when Hash
+                host = host.clone
+                host[:scheme] ||= (host[:scheme] || host[:protocol] || "http").to_s
+                if host[:scheme] == 'http'
+                  URI::HTTP.build(host)
+                elsif scheme == 'https'
+                  URI::HTTPS.build(host)
+                else
+                  raise ArgumentError, "Unrecognized scheme for host #{host}"
+                end
               else
-                raise ArgumentError.new "Method #{method} not supported"
+                raise "Host parameter #{host} is not valid! Try something like 'http://localhost:9200'!"
+            end
+          end
+
+          def setup_sniffing!
+            if options[:sniffing]
+              # We don't support sniffers that aren't threadsafe with timers here!
+              sniffer_class = options[:sniffer_class] ? options[:sniffer_class] : ::Elasticsearch::Transport::Transport::HTTP::Manticore::ManticoreSniffer
+              raise ArgumentError, "Sniffer class #{sniffer_class} must be a ManticoreSniffer!" if sniffer_class.nil? || !sniffer_class.ancestors.include?(::Elasticsearch::Transport::Transport::HTTP::Manticore::ManticoreSniffer)
+              @sniffer = sniffer_class.new(self, logger)
+              @sniffer.sniff_every(options[:sniffer_delay] || 5) do |urls|
+                logger.info("Will update internal host pool with #{urls.inspect}")
+                @pool.update_urls(urls)
               end
-              Response.new resp.code, resp.read_body, resp.headers
             end
           end
 
-          # Builds and returns a collection of connections.
-          # Each connection is a Manticore::Client
-          #
-          # @return [Connections::Collection]
-          #
-          def __build_connections
-            @request_options = {}
+          # Sniff (if enabled) to get the newest list of hosts
+          # then attempt to resurrect any dead URLs
+          def reload_connections!
+            if options[:sniffing]
+              @pool.update_urls(@sniffer.hosts)
+            end
+            @pool.resurrect_dead!
+          end
+
+          def perform_request(method, path, params={}, body=nil)
+            with_request_retries do
+              body = __convert_to_json(body) if body
+              enriching_response(method, path, params, body) do
+                url, response = @pool.perform_request(method, path, params, body)
 
-            if options.key?(:headers)
-              @request_options[:headers] = options[:headers]
+                # Raise an exception so we can catch it for `retry_on_status`
+                __raise_transport_error(response) if response.status.to_i >= 300 && @retry_on_status.include?(response.status.to_i)
+                [url, response]
```

Comment:
The return value is different from `Base#perform_request` here, which returns a `Response` instance, or am I reading the code incorrectly?